### PR TITLE
[while_loop][inductor] fix aliased inputs by cloning

### PR DESCRIPTION
**Pull request by ydwu4**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #160483
* #160467
* #160670
* #160669
* __->__ #160668
* #160374
* #160548

[fx_graph_cse](https://github.com/pytorch/pytorch/blob/main/torch/_functorch/compile_utils.py#L46) is executed in min_cut partitioner which accidentally creates the aliasing for empty buffers and we could see the following graph node for joint graph with cmd: "pytest test/functorch/test_control_flow.py -k test_scan_multiple_layers_gradient_layers_2_device_cpu"
```python
while_loop = torch.ops.higher_order.while_loop(while_loop_cond_graph_0_0, while_loop_body_graph_0_0, (full_default_4, empty_strided_default, full_default_2, full_default_3, full_default_2, full_default_3, full_default, full_default, rev, rev_1, rev_2, rev_3), (primals_4, primals_5, primals_6, primals_7));
```

Notice the operands sequence **"full_default_2, full_default_3, full_default_2, full_default_3, full_default, full_default"**, which indicates the gradient of different layers now sharing the same buffer, which create silent incorrectness.

Fixes https://github.com/pytorch/pytorch/pull/158168.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -8508,6 +8508,35 @@ def __init__(
...
+            return carried_inputs
+
+        # Import clone from lowering module
+        from .lowering import clone
+
+        # Unwrap views to get the underlying buffers for comparison
+        unwrapped_buffers = [
+            buffer.unwrap_view() if isinstance(buffer, ReinterpretView) else buffer
+            for buffer in carried_inputs
+        ]
```
Comment (zou3519) [inline torch/_inductor/ir.py:8528]: Is this legit?

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -8508,6 +8508,35 @@ def __init__(
         self.name = V.graph.register_buffer(self)
         V.graph.register_operation(self)
 
+    @staticmethod
+    def _clone_aliased_inputs(carried_inputs: Sequence[IRNode]) -> Sequence[IRNode]:
```
Comment (zou3519) [inline torch/_inductor/ir.py:8517]: Can you add a test that directly invokes this logic?

Comment (zou3519) [inline torch/_inductor/ir.py:8517]: Also an example of the dce thing here, in a docblock would help.

File: torch/_inductor/ir.py (hunk 1 above)
Comment (ydwu4) [reply torch/_inductor/ir.py:8528]: Yeah, this follows the _has_aliased_buffers implementation, which is the prior art of detecting aliasing in inductor.

File: torch/_inductor/ir.py (hunk 2 above)
Comment (ydwu4) [reply torch/_inductor/ir.py:8517]: Added a scan test. Can only trigger the cse in partitioner. So the following inference code cannot trigger cse:
```python
import torch
def f(x):
    a = torch.zeros_like(x)
    b = torch.zeros_like(x)
    def cond_fn(it, a, b):
        return it < 2

    def body_fn(it, a, b):
        return it + 1, a + 1, b + 1

    it, new_a, new_b = torch._higher_order_ops.while_loop(cond_fn, body_fn, (torch.tensor(0, dtype=torch.int64), a, b))
    return new_a + new_b + it

out = torch.compile(f)(torch.randn(3, 4, requires_grad=True))
```

**Review by zou3519 — APPROVED** [review]
